modules/dnn/src/dnn.cpp

```diff
@@ -208,7 +208,8 @@ class BackendRegistry
         if (cv::ocl::useOpenCL())
         {
             backends.push_back(std::make_pair(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL));
-            backends.push_back(std::make_pair(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16));
+            if (cuda4dnn::doesDeviceSupportFP16())
+                backends.push_back(std::make_pair(DNN_BACKEND_OPENCV, DNN_TARGET_OPENCL_FP16));
```

Comment:
1. There is no CUDNN support on 3.4 branch.
2. You filtering OpenCL not CUDA. Check should be different: https://github.com/opencv/opencv/blob/3.4.17/modules/core/include/opencv2/core/ocl.hpp#L124

BTW, do you aware of this check with warning?

https://github.com/opencv/opencv/blob/4801526f354057fecde2d70a52c69974aa5ee53f/modules/dnn/src/dnn.cpp#L1461-L1468

Reply:
gah! I'm sorry I've mixed both tests without noticing. I'm aware of the warning check, but it would make more sense not to insert cuda_fp16 in the list of available backends since we can detect it's not compatible.